backends/aptcc/apt-intf.cpp

```diff
@@ -2156,6 +2156,16 @@ bool AptIntf::runTransaction(const PkgList &install, const PkgList &remove, cons
         const bool preserveAuto;
     };
 
+    // Calculate existing garbage before the transaction
+    PkgList initial_garbage;
+    if (autoremove) {
+        for (pkgCache::PkgIterator pkg = (*m_cache)->PkgBegin(); ! pkg.end(); ++pkg) {
+            const pkgCache::VerIterator &ver = m_cache->findCandidateVer(pkg);
+            if (!ver.end() && m_cache->isGarbage(pkg))
+                initial_garbage.push_back(ver);
```

Comment:
Building a list of versions seems rather suboptimal given that we are talking about packages, and don't need the versions - it makes it fragile in case candidate versions change during the transaction - we'd not find the garbage anymore.

It's also a vector, which, for large counts of garbage, would make the code below quite slow. My suggestion would be to use an std::unordered_set<pkgCache::Package*>, that should work reliably. You can then just pass pkgCache::PkgIterators (they behave like Package pointers) to insert() and find().